### ISSUE TITLE
Feature request: onTimeUpdate callback

### ISSUE THREAD

**Issue by gonzaloarca** [regosen/Gapless-5#48]
I noticed that one of the biggest callbacks that's missing in this API is the onTimeUpdate callback, similar to [the one that the HTML5 Audio Element](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/timeupdate_event) has. 

Personally, I'd like to use it in order to update state variables used for displaying the player's current position in real time. For the mean time, is there any other alternative that could help me achieve this?

Thanks!

**Comment by regosen** [comment]
That's a good idea!  I could have the Gapless5 player trigger the callback with the value for `getPosition()`  Let me look into this.

**Comment by regosen** [comment]
I just took a stab at it as part of this PR, let me know what you think!  https://github.com/regosen/Gapless-5/pull/47

**Comment by gonzaloarca** [comment]
Awesome! Really appreciate it!

I just tried it out and it looks like the callback function is not being called, I used it like this, in React.

```javascript
gaplessPlayer.ontimeupdate = currentPosition => console.log("ontimeupdate", currentPosition);
``` 

and I cannot see any logs in the console as I click play and the audio effectively starts.

**Comment by regosen** [comment]
Weird, I just tried your line of code as well and it's working for me:
```
gapless5/:164 ontimeupdate 1
gapless5/:164 ontimeupdate 29
gapless5/:164 ontimeupdate 62
...
```
By default Gapless5 doesn't log much to the console, but you can pass `logLevel: LogLevel.Debug` into the constructor options if you want more verbosity.  (I just pushed a fix for types, so if you're using TypeScript you'll want the latest of the PR branch again)

**Comment by regosen** [comment]
As a sanity check, instead of defining `ontimeupdate` can you try the following?
```
console.log(gaplessPlayer.ontimeupdate);
```
Then your JS console should show this:
```
(current_track_time) => {}
```
If instead it prints `undefined` then you're not using the feature branch version of the player.

**Comment by gonzaloarca** [comment]
Alright, I just did that and the console shows
```
currentPosition => console.log("ontimeupdate", currentPosition)
```

**Comment by gonzaloarca** [comment]
Also, completely off topic: have you ever used this library in a React project? I'm trying to figure out what's the best way to maintain state for the UI, in order to reflect the Gapless5's internal state. Using the useState hook with the Gapless5 object (i.e. useState<Gapless5>) doesn't seem too convenient, since the entire object needs to be copied over into a separate object for changes to be detected in the first place (AFAIK). And useRef isn't good for reflecting state changes in the UI. 

Should I just wrap every UI action into a wrapper function and maintain the state in a separate object?

**Comment by regosen** [comment]
> Alright, I just did that and the console shows

What I mean is to call my line _instead_ of defining `gaplessPlayer.ontimeupdate` yourself. That will allow me to check if you have a version of Gapless5 that defines the callback.

Good question about reflecting Gapless 5's internal state.  Depending on your UI you could just access the few things that matter to the UI (such as the current track and the current track's position).  To make it play nicely with React you may need to call accessors on every tick (e.g. getTrack(), isPlaying(), getPosition()), but those calls are cheap and you could probably pass the returned values into your components and have your hooks or contexts memoize those values as dependencies.

**Comment by gonzaloarca** [comment]
Good call, it was `undefined` as you said. Turns out running `yarn add regosen/Gapless-5#pull/47/head` or `yarn upgrade ...` gets the new type definitions but not the actual new implementation. Weird. Not used to using Node packages straight from a repo. Removing and re-adding the package did the trick.

I did notice that the callback still fires even while it's paused. Is that desirable? I do recall the HTML5 Audio Element's callback only being fired while playing and on play and pause events.

**Comment by regosen** [comment]
Yes, I'm having the callback fire on each tick.  Does the HTML5 Audio Element version only fire the callback when the position has changed?  If so, I can change this to do something similar.

**Comment by regosen** [comment]
Ok, I just pushed a change where the time update doesn't fire while it's not playing

**Comment by gonzaloarca** [comment]
It's still firing after pausing for me, and I made sure to install from the latest commit in the PR

**Comment by regosen** [comment]
That is odd, ok I just pushed another change where I only fire a callback when the time and index have changed.  ontimeupdate now provides the index of the current track as a second parameter, so you'll know you have the update if you receive the second parameter.  (If the second parameter is undefined then you still have an old version)

**Comment by gonzaloarca** [comment]
Well, for some reason the steps I did with yarn stopped working. Tried some more tinkering and finally got the new version to work. Everything's fine, both the `currentLength` and the `ontimeupdate` issues. 

Seriously appreciate you taking the time to implement my suggestions! 

I'm currently migrating from using exclusively the HTML5 Audio Element to this library for playlist playback, exclusively for its gapless feature. I'll probably leave some more suggestions that could make Gapless-5 feel more like the native APIs.

Cheers!
